src/history/base.rs

```diff
@@ -1,8 +1,10 @@
 use std::collections::vec_deque::Iter;
 
+use crate::line_buffer::LineBuffer;
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum HistoryNavigationQuery {
-    Normal,
+    Normal(LineBuffer),
```

Comment:
Do we really want the HistoryNavigation to be aware of `LineBuffer`? in my opinion `HistoryNavigation` stuff should be able to exist on it's own without any dependency on anything related to Reedline. In a sense that it is an independent module that any line editor can use.

Reply:
Sub-crate level modularity is currently not high on the list of design requirements. JT and I talked about this being probably the most elegant way to get the `bash` style recovery of the previously typed line buffer (if your cursor is not at the end for prefix search). This way we avoid additional state in the engine that might get out of sync. Also like this, we have a single authoritative line buffer in the `Reedline` engine all the other plugins can rely on. We probably don't have unlimited mental "state budget" for the engine, so we cheated that into this enum.

Reply:
> Sub-crate level modularity is currently not high on the list of design requirements.

Ok, if by sub-carte level modulatiry you mean modular design inside this crate then won't you say that this has the potential to hinder the development later down the road?
If you were hinting about history as a separate create then I did not mean it that way, I was just using that as an example on how I think we should envision the history module API.

Although I would say that I am not really sure what is the exact problem we are trying to solve here and if there even is a better way to do so.